iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/SpecificVersionTagCheck.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * SonarQube IaC Plugin
+ * Copyright (C) 2021-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.iac.docker.checks;
+
+import java.util.List;
+import org.sonar.check.Rule;
+import org.sonar.iac.common.api.checks.CheckContext;
+import org.sonar.iac.common.api.checks.IacCheck;
+import org.sonar.iac.common.api.checks.InitContext;
+import org.sonar.iac.docker.symbols.ArgumentResolution;
+import org.sonar.iac.docker.tree.api.Alias;
+import org.sonar.iac.docker.tree.api.Body;
+import org.sonar.iac.docker.tree.api.DockerImage;
+import org.sonar.iac.docker.tree.api.FromInstruction;
+
+@Rule(key = "S6596")
+public class SpecificVersionTagCheck implements IacCheck {
+
+  private static final String MESSAGE = "Use a specific version tag for the image.";
+
+  @Override
+  public void initialize(InitContext init) {
+    init.register(FromInstruction.class, SpecificVersionTagCheck::checkFromInstruction);
+  }
+
+  private static void checkFromInstruction(CheckContext ctx, FromInstruction fromInstruction) {
+    ArgumentResolution resolvedImage = ArgumentResolution.of(fromInstruction.image());
+    if (resolvedImage.isUnresolved()) {
+      return;
+    }
+    String fullImageName = resolvedImage.value();
+
+    if (hasSensitiveVersionTag(fullImageName) && !imageNameIsUsedAsAliasInPreviousImages(fromInstruction, fullImageName)) {
+      ctx.reportIssue(fromInstruction.image().textRange(), MESSAGE);
+    }
+  }
+
+  private static boolean hasSensitiveVersionTag(String fullImageName) {
+    if (fullImageName.contains("@")) {
+      return false;
+    } else if (fullImageName.contains(":")) {
+      // raise an issue if the version tag is "latest"
+      String[] splitImageName = fullImageName.split(":");
+      return splitImageName.length >= 2 && splitImageName[1].equals("latest");
+    } else {
+      // no version tag specified, docker assumes "latest"
+      return true;
+    }
+  }
+
+  private static boolean imageNameIsUsedAsAliasInPreviousImages(FromInstruction currentFromInstruction, String currentImageName) {
```

Comment:
Not necessarily a big issue here because I don't expect dockerfile to contain a ton of docker image definitions, but this approach has an n^2 complexity, as every time you add a docker image, this will perform an additional iteration looping itself on all previously defined image.
A simpler way would be to have a local cache in this class that retains visited docker images (or even just their alias), but we need to carefully clean this cache or make it available per dockerfile.
I honestly don't have an exact idea about how to implement this or if a similar mechanism already exist.

Reply:
Yeah, I'm not sure what would be the best way either.
Normally we don't want to save the state in between checks, so I figured because we don't expect a lot of FROM instructions in a docker file, that this is save.